File: python/y2017/d03/aoc_2017_03.py
import itertools

from common import Helpers, Puzzle, timer


class Spiral:
    def __init__(self, pos=(0, 0)):
        self.pos = pos
        self.grid = {self.pos: 1}
        self.deltas = self.get_deltas()

    def get_deltas(self):
        return set(itertools.product((-1, 0, 1), (-1, 0, 1))).difference(set({(0, 0)}))

    def next_step(self):
        x, y = self.pos

        if (x > 0) and (abs(x) > abs(y)):
            y += 1
        elif (x < 0) and ((abs(x) > abs(y)) or (x == -y)):
            y -= 1
        elif (y > 0) and ((abs(y) > abs(x)) or (x == y)):
            x -= 1
        else:
            x += 1

        self.grid[(x, y)] = self.grid[self.pos] + 1
        self.pos = (x, y)

        return self.grid[self.pos]

    def get_new_value(self):
        res = 0
        for dx, dy in self.deltas:
            if s := self.grid.get((self.pos[0] + dx, self.pos[1] + dy)):
                res += s
        self.grid[self.pos] = res
        return res


class Day03(Puzzle, year=2017, day=3):
    @staticmethod
    def clean_input(data: list[str]) -> list[str]:
        return data

    @timer(part=1)
    def solve_part_1(self) -> int:
        s = Spiral()
        target = int(self.data[0])

        for _ in range(2, target + 1):
            s.next_step()

        return Helpers.manhattan(s.pos)

    @timer(part=2)
    def solve_part_2(self):
        """Solution for part 2"""
        s = Spiral()
        target = int(self.data[0])

        while True:
            s.next_step()
            if s.get_new_value() > target:
                break

        return s.grid[s.pos]

File: C++/2017/03/aoc_2017_03.cpp
#include "aoc_io.hpp"
#include "aoc_timer.hpp"
#include "aoc_types.hpp"
#include <numeric>
#include <ranges>
#include <string>
#include <utility>
#include <vector>

namespace aoc_2017_03 {

class Spiral {
  private:
    aoc::Grid           grid;
    std::pair<int, int> position;

  public:
    explicit Spiral() {
        position = {0, 0};
        grid.insert(position, 1);
    }
    void step() {
        int x = position.first;
        int y = position.second;

        if ((x > 0) && (std::abs(x) > abs(y))) {
            y += 1;
        } else if ((x < 0) && ((abs(x) > abs(y)) || (x == -y))) {
            y -= 1;
        } else if ((y > 0) && ((abs(y) > abs(x)) || (x == y))) {
            x -= 1;
        } else {
            x += 1;
        }

        position = {x, y};
    }

    unsigned grid_value() {
        return grid.get_value(position);
    }

    void update_value() {
        // clang-format off
        auto res = grid.deltas
            | std::views::transform([this](const auto &p) { return grid.get_value({position.first + p.first, position.second + p.second}); })
            | std::views::common;
        // clang-format on

        unsigned result = std::reduce(res.begin(), res.end());
        grid.insert(position, result);
    }

    unsigned distance() {
        return abs(position.first) + abs(position.second);
    }
};

int solve_part_1(const std::vector<std::string> &input) {
    unsigned target = std::stoul(input[0]);
    Spiral   s;

    for (unsigned i = 1; i < target; i++) {
        s.step();
    }

    return s.distance();
}

int solve_part_2(const std::vector<std::string> &input) {
    unsigned target = std::stoul(input.front());
    Spiral   s;
    unsigned result;

    while (s.grid_value() <= target) {
        s.step();
        s.update_value();
    }

    return s.grid_value();
}

int solve_all(const std::vector<std::string> &input) {
    aoc::timer(1, solve_part_1, input);
    aoc::timer(2, solve_part_2, input);

    return 0;
}
} // namespace aoc_2017_03

int main(int argc, char **argv) {
    std::string filename;
    const int   year = 2017;
    const int   day = 3;

    std::vector<std::string> data;

    if (argc > 1) {
        filename = argv[1];
    } else {
        filename = "input.txt";
    }

    data = aoc::io::get_input_list<std::string>(filename, year, day);

    aoc::io::header(year, day);
    aoc::timer(0, aoc_2017_03::solve_all, data, false);

    return 0;
}
